File: frontend/src/lib/roles.ts
// Mirrors backend pkg/domain/roles.go. Maps usetype → display name and the
// single role each tier may create.
export const USETYPE = {
  SUPER_DUPER_ADMIN: 0,
  COMPANY: 11,
  ADMIN: 10,
  SUB_ADMIN: 9,
  SUPER_MASTER: 8,
  MASTER: 1,
  DEALER: 2,
  PLAYER: 3,
  HELPER: 55,
} as const

const NAMES: Record<number, string> = {
  0: 'Super Duper Admin',
  11: 'Company',
  10: 'Admin',
  9: 'Sub Admin',
  8: 'Super Master',
  1: 'Master',
  2: 'Dealer',
  3: 'End User (Player)',
  55: 'Helper',
}

export const roleName = (usetype: number): string => NAMES[usetype] ?? 'Unknown'

// Reference-style labels for the create selector / role sections.
export const ROLE_LABEL: Record<number, string> = {
  0: 'Super Duper Admin', 11: 'Company', 10: 'Admin', 9: 'Sub Admin',
  8: 'Super Stockist', 1: 'Stockist', 2: 'Dealer', 3: 'User', 55: 'Helper',
}

// Hierarchy rank top→bottom (mirrors backend domain.rank).
const RANK: Record<number, number> = { 0: 0, 11: 1, 10: 2, 9: 3, 8: 4, 1: 5, 2: 6, 3: 7 }
const DOWNLINE = [10, 9, 8, 1, 2, 3] // Admin … User

// creatableRoles: SDA → Company only; Company…Dealer → any tier strictly below
// them; User → none. Mirrors backend Usetype.CreatableRoles.
export const creatableRoles = (usetype: number): number[] => {
  if (usetype === 0) return [11]
  const r = RANK[usetype]
  if (r === undefined || usetype === 3) return []
  return DOWNLINE.filter((t) => RANK[t] > r)
}

// childRoleName — the immediate next tier (default / labels).
export const childRoleName = (usetype: number): string | null => {
  const child = creatableRoles(usetype)[0]
  return child === undefined ? null : NAMES[child]
}

File: backend/pkg/domain/roles.go
// Package domain holds core types shared across services.
package domain

// Usetype is the legacy numeric role code, kept for data compatibility.
type Usetype int

const (
	SuperDuperAdmin Usetype = 0
	Company         Usetype = 11
	Admin           Usetype = 10
	SubAdmin        Usetype = 9
	SuperMaster     Usetype = 8
	Master          Usetype = 1
	Dealer          Usetype = 2
	Player          Usetype = 3
	Helper          Usetype = 55
)

var roleNames = map[Usetype]string{
	SuperDuperAdmin: "Super Duper Admin",
	Company:         "Company",
	Admin:           "Admin",
	SubAdmin:        "Sub Admin",
	SuperMaster:     "Super Master",
	Master:          "Master",
	Dealer:          "Dealer",
	Player:          "End User (Player)",
	Helper:          "Helper",
}

// rank orders the tiers top→bottom. A management tier may create ANY tier below
// it (Company → Admin … User), except the Super Duper Admin who creates Company
// (whitelabel) only, and the User who creates nothing.
var rank = map[Usetype]int{
	SuperDuperAdmin: 0,
	Company:         1,
	Admin:           2,
	SubAdmin:        3,
	SuperMaster:     4, // Super Stockist
	Master:          5, // Stockist
	Dealer:          6,
	Player:          7, // User
}

// downlineOrder lists the creatable tiers below Company, top→bottom.
var downlineOrder = []Usetype{Admin, SubAdmin, SuperMaster, Master, Dealer, Player}

// Name returns the display name for a usetype.
func (u Usetype) Name() string {
	if n, ok := roleNames[u]; ok {
		return n
	}
	return "Unknown"
}

// CanCreate reports whether role u may create role target.
//   - Super Duper Admin → Company only.
//   - Company … Dealer  → any tier strictly below them (but never another Company).
//   - User              → nothing.
func (u Usetype) CanCreate(target Usetype) bool {
	if u == SuperDuperAdmin {
		return target == Company
	}
	ur, ok := rank[u]
	tr, ok2 := rank[target]
	if !ok || !ok2 || target == Helper || target == Company {
		return false
	}
	return u != Player && tr > ur
}

// CreatableRoles returns every tier this role may create, top→bottom.
func (u Usetype) CreatableRoles() []Usetype {
	if u == SuperDuperAdmin {
		return []Usetype{Company}
	}
	var out []Usetype
	for _, t := range downlineOrder {
		if u.CanCreate(t) {
			out = append(out, t)
		}
	}
	return out
}

// Creates returns the immediate next tier below this role (the default when no
// specific target is requested). For the SDA that is Company.
func (u Usetype) Creates() (Usetype, bool) {
	roles := u.CreatableRoles()
	if len(roles) == 0 {
		return 0, false
	}
	return roles[0], true
}
